Added cleanup script for hanging Chrome processes and temp files

cedricroijakkers opened the pull request:
## **User description**
### Description
Added a script (running as a daemon of sorts under supervisord) to clean up files and processes left there by the Chrome browser.

### Motivation and Context
We've been running a Selenium grid for quite some time, using it to execute all kinds of automated monitoring tasks on websites, both running in the internal network and on the internet.

Since we're using the Chrome browser exclusively and our grids are long-running, we've noticed that after some time two types of items remain in the containers (consuming PIDs and disk space):

- Hanging Chrome browser processes. Most of our jobs run for a few minutes, yet after many hours, some Chrome processes remain active. 
- Temporary files of the format `/tmp/.com.google.Chrome.*`, after some time these consume quite a lot of disk space, blowing up the container file system.

I have created a shell script that cleans up these two items which runs once an hour. Temporary files on disk will be removed immediately, Chrome processes after they have been active for over 2 hours. This is what we are doing in our setup using an ugly hack by doing it as part of the Docker healthcheck, and it solved all our capacity problems.

The script has only been added to the Chrome node image, since I cannot tell if the same issue is also present on other browsers. Even if it is, the fix will be different since the browser is different, so the change is not applicable there.

A new script has been added under `/opt/bin/chrome-cleanup.sh` in the container, and a supervisord service is added using configuration file `/etc/supervisor/conf.d/chrome-cleanup.conf`.

### Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
- [X] I have read the [contributing](https://selenium.dev/documentation/en/contributing/) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.


___

## **Type**
Bug fix, Enhancement


___

## **Description**
- Adds a new shell script (`chrome-cleanup.sh`) to clean up hanging Chrome processes and temporary files in the NodeChrome Docker image.
- The cleanup script targets Chrome temp files older than 1 day and processes older than 2 hours for termination.
- The script is integrated into the Docker image via the Dockerfile and is managed by supervisord (`chrome-cleanup.conf`), ensuring it runs continuously as a daemon.
- This change addresses issues with lingering Chrome processes and bloated temp directories in long-running Selenium grid containers.



___



## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>chrome-cleanup.sh</strong><dd><code>Add Chrome Cleanup Daemon Script</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
NodeChrome/chrome-cleanup.sh

<li>Adds a script to delete Chrome temp files and kill hanging Chrome <br>processes.<br> <li> Temp files older than 1 day and processes older than 2 hours are <br>targeted.<br> <li> Script runs in an infinite loop, with a cleanup cycle every hour.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/docker-selenium/pull/2173/files#diff-a10eaaf8e38b1f2d5c9956ee69d1ec150ec58db0b901754d98ecb7f718701521">+31/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Dockerfile</strong><dd><code>Integrate Chrome Cleanup Script into Docker Image</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
NodeChrome/Dockerfile

<li>Adds the Chrome cleanup script and supervisord configuration file to <br>the Docker image.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/docker-selenium/pull/2173/files#diff-472dbd85c9d65177f8d5dfbfdd9c8ee2ec7f54201808b8cd118c5ef399246cd2">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>chrome-cleanup.conf</strong><dd><code>Supervisord Configuration for Chrome Cleanup</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
NodeChrome/chrome-cleanup.conf

<li>Adds a supervisord configuration for the Chrome cleanup script.<br> <li> Configures logging, autostart, and restart policies for the cleanup <br>script.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/docker-selenium/pull/2173/files#diff-6cbfe56ab3510b06fc76a611a1cdb0e9d22b45f208f0ad17f060fb67763a9bf6">+20/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

cedricroijakkers committed:
Added cleanup script for hanging Chrome processes and temp files

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/SeleniumHQ/docker-selenium?pullRequest=2173) <br/>All committers have signed the CLA.

diemol reviewed (commented):
This is probably a good idea. Do you have time to add the same to Edge and Firefox?

cedricroijakkers commented:
> This is probably a good idea. Do you have time to add the same to Edge and Firefox?

I would have to investigate if and where those browsers save their files, but yes. I will check somewhere next week.

VietND96 commented:
Do you think the older time in seconds (default 1200) and the daemon interval in seconds (default 3600) are able to set different values via an ENV var?

cedricroijakkers commented:
> Do you think the older time in seconds (default 1200) and the daemon interval in seconds (default 3600) are able to set different values via an ENV var?

While creating this PR I was actually thinking that too, yes. Maybe even create an ENV var to disable the cleanup alltogether? I will have a look and update the PR.

VietND96 commented:
Yes, ENV var to enable/disable is also fine. I see supervisord is used to start the script, so you can disable it in .conf something like

https://github.com/SeleniumHQ/docker-selenium/blob/db9159801c087b75031fb5ecacd368fdae54f3d4/Video/supervisord.conf#L40-L42

cedricroijakkers committed:
- Added cleanup scripts for Edge and Firefox
- Allowed cleanup to be enabled and/or configured via environment variables
- Updated documentation with new options

cedricroijakkers commented:
I've updated the PR with your comments:

- Cleanup script added for Firefox and Edge (and cleaned up code a little bit)
- Added env vars to enable/disable the cleanup, and when enabled configure timings and parameters
- Added new environment vars to the documentation

VietND96 commented:
LGTM! How about you, @diemol ?

diemol commented on NodeChrome/Dockerfile:
```diff
@@ -53,6 +53,16 @@ RUN if [ ! -z "$CHROME_DRIVER_VERSION" ]; \
+#============================================
+# Chrome cleanup script and supervisord file
+#============================================
+ENV SE_ENABLE_BROWSER_LEFTOVERS_CLEANUP true
+ENV SE_BROWSER_LEFTOVERS_INTERVAL_SECS 3600
+ENV SE_BROWSER_LEFTOVERS_PROCESSES_SECS 1200
+ENV SE_BROWSER_LEFTOVERS_TEMPFILES_DAYS 1
```
Maybe these can be placed in `NodeBase`? We can avoid some duplication.

diemol reviewed (commented):
Should we leave `SE_ENABLE_BROWSER_LEFTOVERS_CLEANUP` enabled by default?

Several folks use these images for web scraping; sessions could last longer than an hour. 

What do you all think?

cedricroijakkers committed:
Switched default cleanup to disabled, updated documentation accordingly

cedricroijakkers committed:
Moved environment variabled to NodeBase to avoid duplication

cedricroijakkers commented:
> Should we leave `SE_ENABLE_BROWSER_LEFTOVERS_CLEANUP` enabled by default?
> 
> Several folks use these images for web scraping; sessions could last longer than an hour.

Fair point. I thought about this as well. Maybe it is better to err on the side of caution, and leave the cleanup disabled by default.

I have changed the default value to `false` and updated the Readme accordingly.

> Maybe these can be placed in `NodeBase`? We can avoid some duplication.

Done!

cedricroijakkers committed:
Fixed wording

diemol reviewed (approved):
Thank you, @cedricroijakkers!

@VietND96 feel free to merge whenever you have time. Thanks!
